New Tab Page & Favicons Refactor

kylehickinson opened the pull request:
<!-- *Thank you for submitting a pull request, your contributions are greatly appreciated!* -->

## Summary of Changes

#### Blanket Refactor Issues
This pull request fixes #2578
This pull request fixes #2579 

#### NTP Issues
This pull request fixes #2380
This pull request fixes #2071
This pull request fixes https://github.com/brave/brave-ios/issues/2036
This pull request fixes https://github.com/brave/brave-ios/issues/2351
This pull request fixes https://github.com/brave/brave-ios/issues/2059
This pull request fixes https://github.com/brave/brave-ios/issues/2516

#### Keyboard Issues
*Note that the "Auto-Open Keyboard" preference has been removed*

This pull request fixes https://github.com/brave/brave-ios/issues/2078
This pull request fixes https://github.com/brave/brave-ios/issues/2073
This pull request fixes https://github.com/brave/brave-ios/issues/2321 (because the pref no longer exists)

#### Favicon Issues
This pull request fixes https://github.com/brave/brave-ios/issues/2051
This pull request fixes https://github.com/brave/brave-ios/issues/1162
This pull request fixes https://github.com/brave/brave-ios/issues/933
This pull request fixes https://github.com/brave/brave-ios/issues/528
This pull request fixes https://github.com/brave/brave-ios/issues/2362
This pull request fixes https://github.com/brave/brave-ios/issues/2485
This pull request fixes https://github.com/brave/brave-ios/issues/2099
This pull request fixes https://github.com/brave/brave-ios/issues/1865
This pull request fixes https://github.com/brave/brave-ios/issues/1030
This pull request fixes https://github.com/brave/brave-ios/issues/2364

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:

- Test New Tab Page related functions:
   - Verify stats appear & function
   - Verify favorites appear and are left-aligned even with less than 4 items
   - Verify show more only appears when there's more than the visible number of favorites on NTP (between 4-6 depending on iPhone vs iPad portrait & landscape)
   - Verify iPhone/iPad landscape vs portrait
   - Verify background images
   - Verify background images source button
   - Verify SI's & SI button
   - Verify CHP's & QR code button
   - Verify private mode callouts
   - Verify themes
   - Verify no-image-bg light theme shows dark-blue bg now and no gradient

- Test all places favicons are seen: Favorites, Bookmarks, Add Bookmark, History, Search, Back/Forward list, Tab Tray, Rewards Panel, Rewards Tipping/Auto-Contribute/Pending Contributions list, etc.
    - Verify custom icons are shown for super-referrers 
    - Verify bundled icons are shown until you visit the page (therefore parsing and downloading the websites up-to-date favicon)
    - Verify websites that only have small favicons display monograms on the favorites overlay & NTP. (i.e. https://eqbank.ca) but display the icon in other places
    - Verify websites that don't have any valid favicon display monograms (e.g. https://kylehickinson.com)

- Test new favorites overlay basic functionality (tap, edit, reorder, scroll, etc.) 

- Test upgrading from older versions that already have favicons downloaded

## Screenshots:

<details>
<summary>Screenshots</summary>

![Simulator Screen Shot - iPhone 11 Pro - 2020-05-25 at 13 41 06](https://user-images.githubusercontent.com/529104/82833912-6b89ed00-9e8d-11ea-9e6f-62200d45867d.png)

![Simulator Screen Shot - iPhone 11 Pro - 2020-05-25 at 13 41 35](https://user-images.githubusercontent.com/529104/82833919-747abe80-9e8d-11ea-8dc6-14243efc84d8.png)

![Simulator Screen Shot - iPhone 11 Pro - 2020-05-25 at 13 44 15](https://user-images.githubusercontent.com/529104/82834024-d4716500-9e8d-11ea-89ae-c853b7b008ed.png)

![Simulator Screen Shot - iPhone 11 Pro - 2020-05-25 at 13 45 17](https://user-images.githubusercontent.com/529104/82834063-f9fe6e80-9e8d-11ea-8354-01b131057246.png)

</details>


## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

kylehickinson committed:
WIP: New Tab Page Refactor

kylehickinson committed:
iOS 12 long-press support

iccub reviewed (commented):
Some comments and questions just by looking at the code, haven't run it yet.

There will be some more involved after custom homepage code is added, one view to add and some logic.
In general the refactor looks good, there's only one hacky solution of aligning the image credits and other 1 item sections I guess

Will do some functional review later, these are just some initial comments from me

kylehickinson committed:
Share favourites count logic for overflow section + small feedback fixes

kylehickinson committed:
WIP: Favorites overlay

kylehickinson committed:
Merge development into new-tab-page-refactor

kylehickinson committed:
Support for custom themes changes & merge fixes

Adds `NewTabPageBackgroundButtonsView` which will now house any static
static elements that may be placed on the NTP such image credit, brand
logos and share via QR code buttons.

kylehickinson committed:
Fix background focal point calculation offsets

kylehickinson committed:
Fix crash if NTP section updates before dataSource is setup

kylehickinson committed:
Don't log warning about Rewards favicon on for local pages

kylehickinson committed:
Anchor favorites overlay & NTP to layout guide

kylehickinson committed:
Add drag & drop reorder to favorites overlay

kylehickinson committed:
Remove animation for hiding NTP;

kylehickinson committed:
Add interactive keyboard dismiss to fav overlay

kylehickinson committed:
Fix height of favs on NTP

kylehickinson committed:
Finish up favourites reordering

kylehickinson committed:
Move NTP to Tab

kylehickinson committed:
Show correct favicon on tab tray for NTP

kylehickinson committed:
Cleanup & remove old FavoritesViewController

kylehickinson committed:
Rename favorites view controller

kylehickinson committed:
Favorites overlay now hides toolbar

kylehickinson committed:
Fix NTP Memory Leak

kylehickinson committed:
Fix bug where an NTP would remain in view hierarchy when switching modes

kylehickinson committed:
Don't reset background image when switching private modes

This is no longer required as NTPs are linked to Tabs and thus do not
need to regenerate background images when switching modes

kylehickinson committed:
Update NTP favicon layout to look better on iPad portrait & multitasking

Allows for 5 icon on iPad portrait 2/3

kylehickinson committed:
Adjust favorite cell pointer interaction

kylehickinson committed:
Add FRC batch updates for favs overlay

kylehickinson committed:
Remove favs data source & old favs tests

Tests were invalid since these controllers were split

kylehickinson committed:
WIP: Favicon fetching improvements

kylehickinson committed:
Merge development into new-tab-page-refactor

kylehickinson committed:
Fix background colors on default favicons

kylehickinson committed:
More favicon fixes

kylehickinson committed:
Fix browser tray animations

kylehickinson committed:
Dump NTP's from Tabs when receiving a memory warning

kylehickinson committed:
Use monogram fallback on smaller images as well (bookmark, history, etc.)

kylehickinson committed:
Rename `NewFaviconFetcher` back to `FaviconFetcher`

kylehickinson committed:
Unify monogram fallback character selection

kylehickinson committed:
Fix UITableViewCell image shenanigans

kylehickinson committed:
Fix #2516: Dismiss NTP SI notification when SI/backgrounds are disabled

kylehickinson committed:
Fix unit tests

kylehickinson committed:
Fix issue that could overwrite FaviconMO's when downloading stored icons

Also fix missed check for private browsing

kylehickinson committed:
Move SI/QR code on landscape to bottom left instead of centered on iPad

kylehickinson committed:
Fix Domain grabbing for add bookmarks UI

kylehickinson committed:
Fix order of favorites after more than 10 favs are added

And also update CV on reorder completing

kylehickinson committed:
Fix reorder favs drag & drop animations

kylehickinson committed:
Merge development into new-tab-page-refactor

kylehickinson committed:
Fix favicons on Rewards UI

kylehickinson committed:
Restrict drag & drop to Brave

kylehickinson committed:
Add whitespace between functions in NTP section providers

iccub reviewed (changes requested):
First pass on this PR, need run this code and think more about high level/architecture stuff

Other things I found:
- in NTP screen when long pressing on a favorite
- in overlay view when long pressing on a favorite
- view debugger showing constraints conflict for monogram icon on overlay screen
- on iOS 12 favorites cell text in overlaymode is dark even in private mode, makes it hard to read
<img width="504" alt="Zrzut ekranu 2020-05-26 o 00 08 52" src="https://user-images.githubusercontent.com/6950387/82846768-c30a7600-9eea-11ea-91f7-552d70f6a72a.png">

- ddg button on iPhone SE overlaps image credits, not sure if we can easily prevent that
<img width="504" alt="Zrzut ekranu 2020-05-26 o 00 08 57" src="https://user-images.githubusercontent.com/6950387/82846814-0533b780-9eeb-11ea-9889-7b7fe7303b22.png">

kylehickinson committed:
Add some documentation & cleanup.

kylehickinson committed:
Adjust the DDG callout top section inset to account for smaller devices

kylehickinson committed:
Remove redundant check for URL on favicon image view extension

kylehickinson committed:
Fix favorites overlay text color on dark themes in iOS 12

kylehickinson committed:
Merge development into new-tab-page-refactor

kylehickinson committed:
Consolidate monogram letter logic

kylehickinson committed:
Consolidate handling bookmark actions logic

kylehickinson committed:
Add additional documentation in favicon fetcher

kylehickinson committed:
Remove code that deletes all Domain's regardless of Bookmark association

iccub reviewed (approved):
Good job, clean refactor and very much needed

kylehickinson committed:
Fix monogram appearing on top of folders when scrolling in Bookmarks

kylehickinson committed:
Ensure monogram favicons are cleared before resetting image

kylehickinson committed:
Fully-clear monogram favicons at start of favicon load

kylehickinson committed:
Merge development into new-tab-page-refactor

kylehickinson committed:
Fix favorites bleeding under safe area (and padding on bg buttons)

kylehickinson committed:
Fix issue where fresh install would not show favourites overflow

kylehickinson committed:
Also pull precomposed apple-touch-icon's during metadata parse

Fix issue where unsized icon would be used instead of sized ones
